hail/src/main/scala/is/hail/expr/ir/agg/Extract.scala

```diff
@@ -90,6 +91,25 @@ case class Aggs(postAggIR: IR, init: IR, seqPerElt: IR, aggs: Array[AggStateSign
   }
 
   def results: IR = ResultOp(0, aggs)
+
+  def getPhysicalAggs(ctx: ExecuteContext, initBindings: Env[PType], seqBindings: Env[PType]): Array[AggStatePhysicalSignature] = {
+    val initsAB = InferPType.newBuilder[InitOp](aggs.length)
+    val seqsAB = InferPType.newBuilder[SeqOp](aggs.length)
+    val init2 = LoweringPipeline.compileLowerer.apply(ctx, init, false).asInstanceOf[IR]
+    val seq2 = LoweringPipeline.compileLowerer.apply(ctx, seqPerElt, false).asInstanceOf[IR]
+    InferPType(init2.noSharing, initBindings, null, inits = initsAB, null)
+    InferPType(seq2.noSharing, seqBindings, null, null, seqs = seqsAB)
+
+    val pSigs = aggs.indices.map { i => InferPType.computePhysicalAgg(aggs(i), initsAB(i), seqsAB(i)) }.toArray
+
+    if (init2 eq init)
```

Comment:
Ok, I think I mostly understand, but please clarify the places where I'm wrong. I think we're performing a lowering pass to generate the same IR that that the caller would eventually execute, to gather all seqOps and initOps, compute the physical agg signature based on this, and then if we had happened to do this with IR that had already been lowered, clear the inferred PTypes because that ir (init, seqPerElt) will be passed over in Compile's call to InferPTypes. 

If so, would it be possible to perform the check first (and if so using those already-inferred isgnatures) to avoid having to re-traverse the ir tree? Also, would it be potentially easier to add a flag on IR to state whether a certain optimization pass had already been applied (much like the Set( ) of restrictions)